src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -7828,6 +7828,419 @@ static void M()
+                //     static int? M7() => throw null;
+                Diagnostic(ErrorCode.ERR_CannotUseManagedTypeInUnmanagedCallersOnly, "int?").WithArguments("int?", "return").WithLocation(24, 12),
+                // (27,20): error CS8894: Cannot use 'int?' as a parameter type on a method attributed with 'UnmanagedCallersOnly.
+                //     static void M8(int? o) {}
+                Diagnostic(ErrorCode.ERR_CannotUseManagedTypeInUnmanagedCallersOnly, "int? o").WithArguments("int?", "parameter").WithLocation(27, 20));
+        }
+
+        [Fact]
+        public void UnmanagedCallersOnlyRequiresUnmanagedTypes_Valid()
+        {
+            var comp = CreateCompilation(new[] { @"
+using System.Runtime.InteropServices;
+class C
+{
+    [UnmanagedCallersOnly]
+    static int M1() => throw null;
+
+    [UnmanagedCallersOnly]
+    static void M2(int o) {}
+
+    [UnmanagedCallersOnly]
+    static T M3<T>() where T : unmanaged => throw null;
+
+    [UnmanagedCallersOnly]
+    static void M4<T>(T t) where T : unmanaged {}
+}
+", UnmanagedCallersOnlyAttribute });
+
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void UnmanagedCallersOnlyRequiresUnmanagedTypes_StructWithGenericUnmanagedType()
+        {
+            var comp = CreateCompilation(new[] { @"
+using System.Runtime.InteropServices;
+public struct S<T> where T : unmanaged
+{
+    public T t;
+}
+class C
+{
+    [UnmanagedCallersOnly]
+    static S<T> M1<T>() where T : unmanaged => throw null;
+
+    [UnmanagedCallersOnly]
+    static void M2<T>(S<T> o) where T : unmanaged {}
+}
+", UnmanagedCallersOnlyAttribute });
+
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void UnmanagedCallersOnlyWithLoopInDefinition()
+        {
+            var comp = CreateCompilation(@"
+#nullable enable
+namespace System.Runtime.InteropServices
+{
+    [AttributeUsage(AttributeTargets.Method, Inherited = false)]
+    public sealed class UnmanagedCallersOnlyAttribute : Attribute
+    {
+        public UnmanagedCallersOnlyAttribute()
+        {
+        }
+
+        public Type[]? CallConvs;
+        public string? EntryPoint;
+
+        [UnmanagedCallersOnly]
+        void M() {}
+    }
+}
+");
+
+            comp.VerifyDiagnostics(
+                // (15,10): error CS8892: `UnmanagedCallersOnly` can only be applied to ordinary static methods or static local functions.
+                //         [UnmanagedCallersOnly]
+                Diagnostic(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, "UnmanagedCallersOnly").WithLocation(15, 10)
+            );
+        }
+
+        [Fact]
+        public void UnmanagedCallersOnlyWithLoopInUsage()
+        {
+            var comp = CreateCompilation(new[] { @"
+using System.Runtime.InteropServices;
+class C
+{
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(C) })]
+    public static void Func() {}
+}
```

Comment:
Consider testing:
```C#
class A
{
    struct B { }
    [UnmanagedCallersOnly(CallConvs = new[] { typeof(B) })]
    static void F() { }
}
```
```C#
class C
{
    [UnmanagedCallersOnly(CallConvs = F())]
    static Type[] F() { }
}
```

Reply:
Consider testing:
```C#
namespace System.Runtime.InteropServices
{
    public sealed class UnmanagedCallersOnlyAttribute : Attribute
    {
        [UnmanagedCallersOnly(CallConvs = new[] { typeof(UnmanagedCallersOnlyAttribute) }]
        public UnmanagedCallersOnlyAttribute() { }
        public Type[] CallConvs;
    }
}
```

---
In reply to: [476674944](https://github.com/dotnet/roslyn/pull/47008#discussion_r476674944) [](ancestors = 476674944)

Reply:
Filed [https://github.com/dotnet/roslyn/issues/47125](https://github.com/dotnet/roslyn/issues/47125) for the general stackoverflow when testing the second one above.

---
In reply to: [476678534](https://github.com/dotnet/roslyn/pull/47008#discussion_r476678534) [](ancestors = 476678534,476674944)